util/codec/bytes.go

```diff
@@ -38,95 +21,94 @@ import (
-// EncodeBytes will append \x00\x01 at the end of the encoded value to
-// indicate the termination.
+const (
+	encGroupSize = 8
+	encMarker    = byte(0xFF)
+	encPad       = byte(0x0)
+)
+
 // EncodeBytes guarantees the encoded value is in ascending order for comparison,
-// The encoded value is >= SmallestNoneNilValue and < InfiniteValue.
+// encoding with the following rule:
+//  [group1][marker1]...[groupN][markerN]
+//  group is 8 bytes slice which is padding with 0.
+//  marker is `0xFF - padding 0 count`
+// For example:
+//   [] -> [0, 0, 0, 0, 0, 0, 0, 0, 247]
+//   [1, 2, 3] -> [1, 2, 3, 0, 0, 0, 0, 0, 250]
+//   [1, 2, 3, 0] -> [1, 2, 3, 0, 0, 0, 0, 0, 251]
+//   [1, 2, 3, 4, 5, 6, 7, 8] -> [1, 2, 3, 4, 5, 6, 7, 8, 255, 0, 0, 0, 0, 0, 0, 0, 0, 247]
+// Refer: https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format
 func EncodeBytes(b []byte, data []byte) []byte {
-	// Allocate more space to avoid unnecessary slice growing
-	bs := reallocBytes(b, len(data)+20)
-	if len(data) > 0 && data[0] == 0xFF {
-		// we must escape 0xFF here to guarantee encoded value < InfiniteValue \xFF\xFF.
-		bs = append(bs, 0xFF, 0x00)
-		data = data[1:]
-	}
-
-	for {
-		// find 0x00 and escape it.
-		i := bytes.IndexByte(data, 0x00)
-		if i == -1 {
-			break
+	// Allocate more space to avoid unnecessary slice growing.
+	// Assume that the byte slice size is about `(len(data) / encGroupSize + 1) * (encGroupSize + 1)` bytes,
+	// that is `(len(data) / 8 + 1) * 9` in our implement.
+	dl := len(data)
+	reallocSize := (dl/encGroupSize + 1) * (encGroupSize + 1)
+	bs := reallocBytes(b, reallocSize)
+	for idx := 0; idx <= dl; idx += encGroupSize {
+		remain := dl - idx
+		padCount := 0
+		if remain >= encGroupSize {
+			bs = append(bs, data[idx:idx+encGroupSize]...)
+		} else {
+			padCount = encGroupSize - remain
+			bs = append(bs, data[idx:]...)
+			bs = append(bs, make([]byte, padCount)...)
 		}
-		bs = append(bs, data[:i]...)
-		bs = append(bs, 0x00, 0xFF)
-		data = data[i+1:]
-	}
-	bs = append(bs, data...)
-	return append(bs, 0x00, 0x01)
-}
-
-// DecodeBytes decodes bytes which is encoded by EncodeBytes before,
-// returns the leftover bytes and decoded value if no error.
-func DecodeBytes(b []byte) ([]byte, []byte, error) {
-	return decodeBytes(b, 0xFF, 0x00, 0x01)
-}
 
-func decodeBytes(b []byte, escapeFirst byte, escape byte, term byte) ([]byte, []byte, error) {
-	if len(b) < 2 {
-		return nil, nil, errors.Errorf("insufficient bytes to decode value")
+		marker := encMarker - byte(padCount)
+		bs = append(bs, marker)
 	}
 
-	var r []byte
+	return bs
+}
 
-	if b[0] == escapeFirst {
-		if b[1] != ^escapeFirst {
-			return nil, nil, errors.Errorf("invalid escape byte, must 0x%x, but 0x%x", ^escapeFirst, b[1])
+func decodeBytes(b []byte, reverse bool) ([]byte, []byte, error) {
+	data := make([]byte, 0, len(b))
+	for {
+		if len(b) < encGroupSize+1 {
+			return nil, nil, errors.New("insufficient bytes to decode value")
```

Comment:
Seems meaningless to log it, for it is in a for loop, the len(b) is not the origin b maybe.